File: bridges/nvda/tests/fakes/document_reader.py
# nvdaMcpBridge tests -- FakeDocumentReader, standing in for the DocumentReader port.
# Copyright (C) 2026 Marlon Brandao de Sousa. GPL-2. See COPYING.txt.
# has_document=False stands for a focus with no document at all, not an empty page.

from __future__ import annotations

from collections.abc import Sequence

from nvdaMcpBridge.domain.entities.document_snapshot import DocumentSnapshot
from nvdaMcpBridge.domain.ports.document_reader import DocumentRead, DocumentReader


class FakeDocumentReader(DocumentReader):
	def __init__(
		self,
		lines: Sequence[str] | None = None,
		*,
		title: str = "",
		has_document: bool = True,
	) -> None:
		self.lines = list(lines or [])
		self.title = title
		self.has_document = has_document
		#: Lines actually rendered, so a test can prove a bound stopped the walk.
		self.offered = 0

	def read(self, snapshot: DocumentSnapshot) -> DocumentRead | None:
		if not self.has_document:
			return None
		for line in self.lines:
			self.offered += 1
			if not snapshot.offer(line):
				break
		return DocumentRead(title=self.title)

File: server/domain/ports/document_reader.go
// screenreader-mcp domain -- the DocumentReader port (the `document` capability).
// Copyright (C) 2026 Marlon Brandao de Sousa. GPL-2. See COPYING.txt.
//
// ROLE: domain port, the `document` capability group.
// IMPLEMENTED BY: adapters/bridge/json_lines_client.go.
// USED BY: the get_document_snapshot tool controller.
// HANDED OUT BY: the handshake, only when the reader announced `document`.
package ports

// SnapshotLine ordinals are absolute, whatever line the read started at.
type SnapshotLine struct {
	Line int
	Text string
}

// DocumentSnapshot.CapturedAt is the reader's own wall clock, in its log's format.
type DocumentSnapshot struct {
	// HasDocument false means no document at all, not an empty one.
	HasDocument bool
	CapturedAt  string
	Title       string
	Lines       []SnapshotLine
	FromLine    int
	ToLine      int
	// TruncatedBy is "none", "maxLines" or "maxChars", never empty.
	TruncatedBy string
}

// DocumentBounds zero means no limit, so the zero value is the whole document.
type DocumentBounds struct {
	FromLine int
	MaxLines int
	MaxChars int
}

type DocumentReader interface {
	Snapshot(bounds DocumentBounds) (DocumentSnapshot, error)
}
